File: semester 4/MPP/JFTS/FestivalPersistance/src/main/java/ro/mpp/repository/DBRepository/TicketRepository.java
package ro.mpp.repository.DBRepository;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import ro.mpp.domain.Show;
import ro.mpp.domain.Ticket;
import ro.mpp.exceptions.TicketModifier;
import ro.mpp.exceptions.ValidatorException;
import ro.mpp.repository.ITicketRepository;
import ro.mpp.utils.Database;
import ro.mpp.validator.ValidatorStrategy;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public class TicketRepository implements ITicketRepository {
    private final Database db;
    private final ShowRepository showRepository;
    private final ValidatorStrategy<Ticket> ticketValidator;
    private static final Logger logger = LogManager.getLogger(TicketRepository.class);

    public TicketRepository(Database db, ShowRepository showRepository, ValidatorStrategy<Ticket> ticketValidator) {
        this.db = db;
        this.showRepository = showRepository;
        this.ticketValidator = ticketValidator;
    }

    @Override
    public Optional<Ticket> incrementSeats(Ticket ticket, int seats) throws TicketModifier {
        logger.info("Increment seats for ticket {} from {} to {}", ticket.getId(), ticket.getNumberOfSeats(), seats);

        String sql = "UPDATE ticket SET number_of_seats = ? WHERE id = ?";
        try {
            Connection conn = db.getConnection();
            try (PreparedStatement ps = conn.prepareStatement(sql)) {
                ps.setInt(1, seats);
                ps.setInt(2, ticket.getId());
                ps.executeUpdate();
            }

            return Optional.of(ticket);
        } catch (SQLException e) {
            logger.error(e.getMessage());
            throw new RuntimeException(e);
        }
    }

    @Override
    public Optional<Ticket> save(Ticket entity) throws ValidatorException {
        logger.info("Validating ticket {}", entity);
        ticketValidator.validate(entity);
        logger.traceEntry("TicketRepository save {}", entity);
        String sql = "INSERT INTO ticket (buyer_name, number_of_seats, purchase_date, show_id) VALUES (?, ?, ?, ?)";

        try {
            Connection conn = db.getConnection();
            try (PreparedStatement ps = conn.prepareStatement(sql, PreparedStatement.RETURN_GENERATED_KEYS)) {

                ps.setString(1, entity.getBuyerName());
                ps.setInt(2, entity.getNumberOfSeats());
                ps.setTimestamp(3, entity.getPurchaseDate());
                ps.setInt(4, entity.getShow().getId());

                ps.executeUpdate();

                try (ResultSet rs = ps.getGeneratedKeys()) {
                    if (rs.next()) {
                        entity.setId(rs.getInt(1));
                    }
                }
                return Optional.of(entity);
            }
        } catch (SQLException e) {
            logger.error("Error while saving ticket", e);
            throw new RuntimeException("Error while saving ticket: " + e.getMessage());
        }
    }

    @Override
    public Optional<Ticket> update(Ticket entity) {
        logger.traceEntry("TicketRepository update {}", entity);
        String sql = "UPDATE ticket SET buyer_name = ?, number_of_seats = ?, purchase_date = ? WHERE id = ?";

        try {
            Connection conn = db.getConnection();
            try (PreparedStatement ps = conn.prepareStatement(sql)) {

                ps.setString(1, entity.getBuyerName());
                ps.setInt(2, entity.getNumberOfSeats());
                ps.setTimestamp(3, entity.getPurchaseDate());
                ps.setInt(4, entity.getShow().getId());
                ps.setInt(5, entity.getId());

                ps.executeUpdate();

                return Optional.of(entity);
            }
        } catch (SQLException e) {
            logger.error("Error while updating ticket", e);
            throw new RuntimeException("Error while updating ticket: " + e.getMessage());
        }
    }

    @Override
    public void delete(Ticket entity) {
        logger.traceEntry("TicketRepository delete {}", entity);
        String sql = "DELETE FROM ticket WHERE id = ?";

        try {
            Connection conn = db.getConnection();
            try (PreparedStatement ps = conn.prepareStatement(sql)) {

                ps.setInt(1, entity.getId());
                ps.executeUpdate();
            }
        } catch (SQLException e) {
            logger.error("Error while deleting ticket", e);
            throw new RuntimeException("Error while deleting ticket: " + e.getMessage());
        }
    }

    @Override
    public Optional<Ticket> find(Integer id) {
        logger.traceEntry("TicketRepository find ticket with id {}", id);
        String sql = "SELECT * FROM ticket WHERE id = ?";

        try {
            Connection conn = db.getConnection();
            try (PreparedStatement ps = conn.prepareStatement(sql)) {
                ps.setInt(1, id);

                try (ResultSet rs = ps.executeQuery()) {
                    if (rs.next()) {
                        Show show = showRepository
                                .find(rs.getInt("show_id"))
                                .orElseThrow(() -> {
                                    logger.error("Error while finding show for ticket with id {}", id);
                                    return new RuntimeException("Show not found");
                                });

                        return Optional.of(new Ticket(rs, show));
                    }
                }
            }
        } catch (SQLException e) {
            logger.error("Error while finding ticket with id {}", id, e);
            throw new RuntimeException("Error while finding ticket: " + e.getMessage());
        }
        return Optional.empty();
    }

    @Override
    public List<Ticket> findAll() {
        logger.traceEntry("TicketRepository findAll");
        String sql = "SELECT * FROM ticket";
        List<Ticket> tickets = new ArrayList<>();

        try {
            Connection conn = db.getConnection();
            try (PreparedStatement ps = conn.prepareStatement(sql);
                 ResultSet rs = ps.executeQuery()) {
                while (rs.next()) {
                    Integer id = rs.getInt("id");
                    Show show = showRepository
                            .find(id)
                            .orElseThrow(() -> {
                                logger.error("Error while finding show with id {}", id);
                                return new RuntimeException("Show not found");
                            });

                    tickets.add(new Ticket(rs, show));
                }
                return tickets;
            }
        } catch (SQLException e) {
            logger.error("Error while finding all tickets", e);
            throw new RuntimeException("Error while finding all tickets: " + e.getMessage());
        }
    }
}

File: semester 4/MPP/C#FTS/Model/Repository/DBRepository/TicketRepository.cs
using C_FTS.Domain;
using C_FTS.Utils;
using log4net;
using System.Data;

namespace C_FTS.Repository.DBRepository
{

    public class TicketModifierException : Exception
    {
        public TicketModifierException(string message) : base(message)
        {
        }
    }
    public class TicketRepository : ITicketRepository
    {
        private static readonly ILog logger = LogManager.GetLogger(typeof(ArtistRepository));
        private Database db = new Database();
        private ShowRepository showRepository = new ShowRepository();

        public void Delete(Ticket entity)
        {
            throw new NotImplementedException();
        }

        public Ticket? Find(int id)
        {
            logger.DebugFormat("Entering Find ticket with id {0}", id);

            IDbConnection conn = db.GetConnection();

            using (IDbCommand cmd = conn.CreateCommand())
            {
                cmd.CommandText = "SELECT * FROM ticket WHERE id = @id";

                IDbDataParameter dp = cmd.CreateParameter();
                dp.ParameterName = "@id";
                dp.Value = id;
                cmd.Parameters.Add(dp);

                using (var dataR = cmd.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        Show show = showRepository.Find(
                            dataR.GetInt32(dataR.GetOrdinal("show_id"))
                        ) ?? throw new RepositoryException("Show not found");

                        Ticket ticket = new(dataR, show);

                        logger.InfoFormat("Exiting Find with {0}", ticket);
                        return ticket;
                    }
                }
            }

            logger.Error("Exiting Find with null");
            return null;
        }

        public List<Ticket> FindAll()
        {
            throw new NotImplementedException();
        }

        public Ticket? IncrementSeats(Ticket ticket, int seats)
        {
            logger.DebugFormat("Increment seats for ticket {0} from {1} to {2}", ticket.Id, ticket.NumberOfSeats, seats);

            IDbConnection conn = db.GetConnection();

            using(IDbCommand cmd = conn.CreateCommand())
            {
                cmd.CommandText = "UPDATE ticket SET number_of_seats = @seats WHERE id = @id";
                IDbDataParameter p1 = cmd.CreateParameter();
                p1.ParameterName = "@seats";
                p1.Value = seats;
                cmd.Parameters.Add(p1);

                IDbDataParameter p2 = cmd.CreateParameter();
                p2.ParameterName = "@id";
                p2.Value = ticket.Id;
                cmd.Parameters.Add(p2);

                try
                {
                    int result = cmd.ExecuteNonQuery();

                    if (result > 0)
                    {
                        return ticket;
                    }

                    logger.Error("Exiting IncrementSeats with null");
                    return null;
                }
                catch (Exception e)
                {
                    logger.Error("Error while IncrementSeats ticket", e);
                    throw new Exception("Error while IncrementSeats ticket: " + e.Message);
                }
            }
        }

        public Ticket? Save(Ticket entity)
        {
            logger.DebugFormat("Entering Save with entity {0}", entity);

            IDbConnection conn = db.GetConnection();

            using (IDbCommand cmd = conn.CreateCommand())
            {
                cmd.CommandText = "INSERT INTO ticket (buyer_name, number_of_seats, purchase_date, show_id) " +
                                  "VALUES (@buyer_name, @number_of_seats, @purchase_date, @show_id); " +
                                  "SELECT last_insert_rowid();";

                IDbDataParameter p1 = cmd.CreateParameter();
                p1.ParameterName = "@buyer_name";
                p1.Value = entity.BuyerName;
                cmd.Parameters.Add(p1);

                IDbDataParameter p2 = cmd.CreateParameter();
                p2.ParameterName = "@number_of_seats";
                p2.Value = entity.NumberOfSeats;
                cmd.Parameters.Add(p2);

                IDbDataParameter p3 = cmd.CreateParameter();
                p3.ParameterName = "@purchase_date";
                p3.Value = entity.PurchaseDate;
                cmd.Parameters.Add(p3);

                IDbDataParameter p4 = cmd.CreateParameter();
                p4.ParameterName = "@show_id";
                p4.Value = entity.Show.Id;
                cmd.Parameters.Add(p4);

                try
                {
                    object result = cmd.ExecuteScalar();
                    if (result != null)
                    {
                        entity.Id = Convert.ToInt32(result);
                    }

                    logger.DebugFormat("Exiting Save with {0}", entity);
                    return entity;
                }
                catch (Exception e)
                {
                    logger.Error("Error while saving ticket", e);
                    throw new Exception("Error while saving ticket: " + e.Message);
                }
            }
        }

        public Ticket? Update(Ticket entity)
        {
            logger.DebugFormat("Entering Update with entity {0}", entity);

            IDbConnection conn = db.GetConnection();

            using (IDbCommand cmd = conn.CreateCommand())
            {
                cmd.CommandText = "UPDATE ticket SET buyer_name = @buyer_name, " +
                                  "number_of_seats = @number_of_seats, " +
                                  "purchase_date = @purchase_date, " +
                                  "show_id = @show_id WHERE id = @id";

                IDbDataParameter p1 = cmd.CreateParameter();
                p1.ParameterName = "@buyer_name";
                p1.Value = entity.BuyerName;
                cmd.Parameters.Add(p1);

                IDbDataParameter p2 = cmd.CreateParameter();
                p2.ParameterName = "@number_of_seats";
                p2.Value = entity.NumberOfSeats;
                cmd.Parameters.Add(p2);

                IDbDataParameter p3 = cmd.CreateParameter();
                p3.ParameterName = "@purchase_date";
                p3.Value = entity.PurchaseDate;
                cmd.Parameters.Add(p3);

                IDbDataParameter p4 = cmd.CreateParameter();
                p4.ParameterName = "@show_id";
                p4.Value = entity.Show.Id;
                cmd.Parameters.Add(p4);

                IDbDataParameter p5 = cmd.CreateParameter();
                p5.ParameterName = "@id";
                p5.Value = entity.Id;
                cmd.Parameters.Add(p5);

                try
                {
                    int result = cmd.ExecuteNonQuery();

                    if (result > 0)
                    {
                        logger.DebugFormat("Exiting Update with {0}", entity);
                        return entity;
                    }

                    logger.Error("Exiting Update with null");
                    return null;
                }
                catch (Exception e)
                {
                    logger.Error("Error while updating ticket", e);
                    throw new Exception("Error while updating ticket: " + e.Message);
                }
            }
        }
    }
}
